draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+
+Listener
+~~~~~~~~
+
+The ``Listener`` service is responsible for receiving messages and enqueuing
+them in the appropriate :ref:`draft/celery-5-high-level-architecture:Messages Backlog`.
+
+During initialization the service starts listening to incoming TCP connections.
+
+The service is only run in case the user opts to run the Publisher in
+:ref:`draft/celery-5-high-level-architecture:Publisher Daemon` mode.
+
+Publisher Health Checks
++++++++++++++++++++++++
+
+The Publisher will perform health checks to ensure that
+the :term:`Message Broker` the user is publishing to is available.
+
+If a health check fails a configured number of times, the relevant
+:term:`Circuit Breaker` is tripped.
+
+Each :term:`Message Broker` Celery supports must provide an implementation for
+the default health checks the Publisher will use for verifying its
+availability for new :term:`messages <Message>`.
+
+Further health checks can be defined by the user.
+These health checks allows the user to avoid publishing tasks if for example
+a 3rd party API endpoint is not available or slow, if the database
+the user stores the results in is available or any other check for that matter.
+
+Publisher Circuit Breakers
+++++++++++++++++++++++++++
+
+Each :ref:`health check <draft/celery-5-high-level-architecture:Health Checks>`
+has it's own Circuit Breaker.
+Once a circuit breaker is tripped, the :term:`messages <Message>` are stored
+in the :ref:`draft/celery-5-high-level-architecture:messages backlog` until
+the health check recovers and the circuit is once again closed.
+
+Router
+------
+
+The Router is a :term:`Message Dispatcher`.
+It is responsible for managing the connection to a :term:`message broker`
+and consuming :term:`messages <Message>` from the :term:`message broker`.
+
+The Router can maintain a connection to a cluster of
+:term:`message brokers <message broker>` or even clusters of
+:term:`message brokers <message broker>`.
+
+Data Sources and Sinks
+++++++++++++++++++++++
+
+Data Sources are a new concept in Celery.
+Data Sinks are a concept which replaces Result Backends.
+
+Data Sinks consume :term:`Document Messages <Document Message>` while Data Sources
+produce them.
+
+Data Sources
+~~~~~~~~~~~~
+
+Data Sources are :ref:`task <draft/celery-5-high-level-architecture:Services>`
+which either listen or poll for incoming data from a data source such as a
+database, a file system or an HTTP(S) endpoint.
+
+These services produce :term:`Document Messages <Document Message>`.
+
+Tasks which are subscribed to Data Sources will receive the raw document
+messages for further processing.
+
+.. admonition:: Example
+
+  We'd like to design a feature which locks Github issues immediately after
+  they are closed.
+
+  Github uses Webhooks to notify us when an issue is closed.
+
+  We set up a Data Source which starts an HTTPS server and expects incoming
+  HTTP requests on an endpoint.
+
+  Whenever a request arrives a :term:`Document Message` is published.
```

Comment:
Could you please elaborate on the use of Data Sources? At least to me it was not clear, why the above use case cannot be accomplished by an Event message. Are data sources meant to provide an easier API for using the configured message broker to publish messages?